ggml-quants.h

```diff
@@ -229,19 +229,19 @@ void dequantize_row_iq2_xxs(const block_iq2_xxs * restrict x, float * restrict y
 void dequantize_row_iq2_xs (const block_iq2_xs  * restrict x, float * restrict y, int k);
 
 // Dot product
-void ggml_vec_dot_q4_0_q8_0(int n, float * restrict s, const void * restrict vx, const void * restrict vy);
-void ggml_vec_dot_q4_1_q8_1(int n, float * restrict s, const void * restrict vx, const void * restrict vy);
-void ggml_vec_dot_q5_0_q8_0(int n, float * restrict s, const void * restrict vx, const void * restrict vy);
-void ggml_vec_dot_q5_1_q8_1(int n, float * restrict s, const void * restrict vx, const void * restrict vy);
-void ggml_vec_dot_q8_0_q8_0(int n, float * restrict s, const void * restrict vx, const void * restrict vy);
+void ggml_vec_dot_q4_0_q8_0(int n, float * restrict s, const void ** restrict vx, const void ** restrict vy, const int nrc);
+void ggml_vec_dot_q4_1_q8_1(int n, float * restrict s, const void ** restrict vx, const void ** restrict vy, const int nrc);
+void ggml_vec_dot_q5_0_q8_0(int n, float * restrict s, const void ** restrict vx, const void ** restrict vy, const int nrc);
+void ggml_vec_dot_q5_1_q8_1(int n, float * restrict s, const void ** restrict vx, const void ** restrict vy, const int nrc);
+void ggml_vec_dot_q8_0_q8_0(int n, float * restrict s, const void ** restrict vx, const void ** restrict vy, const int nrc);
```

Comment:
I'm not convinced if this API is desirable - it requires preparing arrays of pointers which seems quite cumbersome

Normally, linear algebra libraries utilize an API of a pointer, number of elements and stride (in bytes or in elements). So I'm thinking that we should probably switch to something like:

```c
void ggml_vec_dot_q4_0_q8_0(int n, float * restrict s, const void * restrict vx, size_t bx, const void * restrict vy, size_t by, int nrc);
```

Note that I'm mostly thinking out loud - not yet sure what is the best way.
It's a big change so we have to consider the options to make this less intrusive

Reply:
I agree. as I mentioned earlier I tried to fit it into the existing interface itself but had changed to array mainly to consider stride. if it's better to add few more ags than arrays, how about we define a tensor attribute structure and pass it across instead of adding one arg for each attribute? this way we can extend it is future for any new functionality. For now the tensor object can just have elements, stride, format type.

Reply:
Hm, adding a new tensor attribute structure would again introduce a lot of boilerplate around calling the dot functions. Adding extra arguments is better in this regard, because we already have the strides from the `struct ggml_tensor`

Reply:
Hi @ggerganov,  the number of elements (bx/by) you mean,  the number of rows/cols and stride is the Kth dimension (MxK: KxN)), right. Just wanted to be sure.

Reply:
Currently, `ggml` stores the strides in number of bytes. So the numbers in `ggml_tensor->nb` are strides in bytes. The dot functions should also accept the row strides in bytes for consistency.

In the future, we will transition to storing the strides in number of elements: https://github.com/ggerganov/ggml/issues/623. But this is not important for now

Reply:
Hi @ggerganov , I have updated the PR. please review and let me know if it can be improved further, especially around the stride calculations. I was able to use the ggml_tensor strides (nb) for src0 and dst tensors, but, I had to arrive at the src1_col stride following the logic used for offset calculations.